Supervise `Endpoint`s

luckysori opened the pull request:
Fix #2821.

luckysori committed:
fix(docs): Fix intra-doc links for supervisor

luckysori committed:
feat(supervisor): Add constructors that accept xtra::Context

Taking the `xtra::Context` from the outside allows callers to use the
actor's `xtra::Address` before the `Supervisor` is instantiated.

luckysori committed:
feat(maker): Supervise Endpoint

This is important because we could run into unforeseen panics.

luckysori committed:
feat(taker): Supervise Endpoint

This is important because we could run into unforeseen panics.

klochowicz reviewed (commented):
I am not sure whether this is a good idea - all of the Endpoint users do not expect the endpoint to be restarted as they keep count of the connections it notifies about during runtime.
 if it gets restarted, we will catch the panic, but we will go straight into undefined behaviour territory. I'm sceptical whether we actually need the additional complexity that would supporting this properly bring with it.

Personally, the best way to deal with this would be: #2824 - there are places in the code which, when causing an error, cause such ripples, it would be easier to just start again instead of trying to restare the balance.

note: the only panic that happened ever from the Endpoint so far was locally, when I tried invalid config values.

klochowicz commented on daemon/src/lib.rs:
```diff
@@ -296,34 +296,45 @@ where
-                identify_listener_actor,
-                libp2p_offer_addr,
-            ),
-            endpoint::Subscribers::new(
```
consider adding a new EndpointRestarted message that invalidates the counts on all the observed connections / listeners

IMHO all the subscribers should be interested in that message, as this break the guarantees given earlier by the endpoint.

luckysori replied:
My proposal is https://github.com/itchysats/itchysats/pull/2839/commits/0d1dd1bdb5cc12e38bfcf51dfc4d9cd43a58adf8.

luckysori committed:
feat: Forget peer if dialing fails because of no connection

This is a simpler and more natural way of dealing with disconnected
peers: if we try to use the connection and the source of truth already
tells us that there is no connection, we can forget about this peer.

da-kami commented on xtra-libp2p-offer/src/current/maker.rs:
```diff
@@ -102,9 +117,8 @@ impl Actor {
-    async fn handle_connection_dropped(&mut self, msg: endpoint::ConnectionDropped) {
-        tracing::trace!("Remove dropped connection: {:?}", msg.peer_id);
-        self.connected_peers.remove(&msg.peer_id);
+    async fn handle_no_connection(&mut self, msg: NoConnection) {
```
🤔 I am not sure I would replace one solution with the other. 
I think if the endpoint already knows that the connection was dropped we should also react to it in the protocols that want to know. 

I would add what you proposed as a resilience fallback on top of the previous solution. Otherwise it is expected to see the warning above, and I don't think that is expected behavior.

Note: Same for all other protocols where this was changed.

luckysori replied:
> I think if the endpoint already knows that the connection was dropped we should also react to it in the protocols that want to know.

I couldn't justify to myself the need for the connection-dropped notification in any of the places where I removed the handlers. It's not like anything dangerous can happen with this stale information. As soon as the actor tries to use the information they will get a `NoConnection` error which is functionally the same as receiving an eager notification.

For the places where I didn't remove the handling of connection-dropped notifications, the reasons are pretty specific:

- I wanted to remove it for the `identify::dialer::Actor`, but we currently have that questionable need to report the number of live connections from there. If we didn't have that I think that actor wouldn't need to know about dropped connections at all.
- The `online_status::Actor` is the only place where I think the subscription is a good idea. Although to be fair it is only used by the taker to keep track of one specific connection (the maker's), even though we will hear about all dropped connections (probably only the maker's, but still). One could argue that we might just want to add a simple query mechanism to the `Endpoint` to actively check if the connection is still alive and remove the dropped-connection notifications altogether!

> Otherwise it is expected to see the warning above, and I don't think that is expected behavior.

Nope, the warning can be (and is) skipped if we don't have a connection to the peer.

da-kami replied:
I am not sure I like this solution, because we now can have different state concerning a peer being connected in different actors. 

This might be resilient for these specific actors, but I think the original idea was that the endpoint is in charge of managing connected peers. So far we always got notified about connected peers from the endpoint, i.e. we passed the state of the endpoint on to subscribers. Now these actors change this state independent of the state in the endpoint. I am not saying this is wrong, but I am not sure this is what we want.

luckysori replied:
> So far we always got notified about connected peers from the endpoint, i.e. we passed the state of the endpoint on to subscribers. Now these actors change this state independent of the state in the endpoint.

But it is not independent of the state in the endpoint! It is the endpoint that tells them about an actor no longer being connected, but only when they ask.

If we want to eagerly synchronise these states then we can reintroduce the notifications, but I don't see the need. In fact, I would like to eventually remove the feature altogether. But that might be too much.

da-kami replied:
> It is the endpoint that tells them about an actor no longer being connected, but only when they ask.

I am not sure I can follow. Until this PR we subscribed to state changes of the endpoint, so the endpoint did notify subscribed protocol actors about peer disconnection. 

> If we want to eagerly synchronise these states then we can reintroduce the notifications, but I don't see the need. 

Fair enough, I think the solution you came up with works, it might just not be as efficient as it could be, because we might still try to open a protocol substream with a peer even if the endpoint already knows that the peer idisconneced. Acceptable I suppose.

> In fact, I would like to eventually remove the feature altogether. But that might be too much.

You mean the feature of protocol actors subscribing to state changes of connect/disconnect in the endpoint? Initially we did not have this feature, but that lead to having to query the endpoint about peer information for all protocols where we need all peers. In the long run it might just be more efficient (and resilient) to change these protocols to be polled by the `taker` instead of the `maker` pushing the information. Then, we also don't need `ping` from `maker` to `taker` anymore.

luckysori committed:
chore(endpoint): Remove Error::BadConnection from public API

Consumers do not care if the connection is bad. Internally, we
immediately drop the connection if it's bad. What consumers care about
is that we no longer have a working connection. Letting them know
allows them to forget the corresponding peer for the time being.

da-kami reviewed (approved):
I still don't see a reason why we should not supervise endpoints, so I still approve. 
I added another comment concerning the state of connected peers now being changed independent of the state in the endpoint. I am not sure this is good, but I think it is more resilient for each actor. We should be able to rely on the specific error here.
